File: lcps-board-analyzer/backend/pipeline/validator.py
"""10-check quality gate — runs before any DB write.

Raises ValidationError with a descriptive message if any check fails.
"""
from __future__ import annotations

import logging

logger = logging.getLogger(__name__)

VALID_CATEGORY_SLUGS = {
    "schools-education", "school-construction", "budget-finance",
    "transportation", "zoning-land-use", "public-safety",
    "policy-governance", "equity-inclusion", "technology",
    "community-parks", "personnel", "general",
}

VALID_URGENCY = {"routine", "notable", "significant"}
MIN_SUMMARY_LEN = 80
MAX_SUMMARY_LEN = 2000
MAX_TITLE_LEN = 300
MAX_DECISIONS = 20
INVALID_PLACEHOLDERS = {"n/a", "none", "unknown", "tbd", ""}


class ValidationError(Exception):
    pass


def validate_item(prompt1: dict, prompt2: dict) -> None:
    """Run all 10 quality checks. Raises ValidationError on first failure."""

    # 1. title is a non-empty string
    title = prompt1.get("title", "")
    if not isinstance(title, str) or not title.strip():
        raise ValidationError("Check 1 failed: 'title' is empty or not a string")

    # 2. title length reasonable
    if len(title) > MAX_TITLE_LEN:
        raise ValidationError(f"Check 2 failed: 'title' exceeds {MAX_TITLE_LEN} chars")

    # 3. summary is a non-empty string
    summary = prompt1.get("summary", "")
    if not isinstance(summary, str) or not summary.strip():
        raise ValidationError("Check 3 failed: 'summary' is empty or not a string")

    # 4. summary meets minimum length
    if len(summary.strip()) < MIN_SUMMARY_LEN:
        raise ValidationError(f"Check 4 failed: 'summary' is too short ({len(summary)} < {MIN_SUMMARY_LEN})")

    # 5. summary does not contain placeholder values
    if summary.strip().lower() in INVALID_PLACEHOLDERS:
        raise ValidationError("Check 5 failed: 'summary' contains invalid placeholder content")

    # 6. decisions is a list (may be empty)
    decisions = prompt1.get("decisions", [])
    if not isinstance(decisions, list):
        raise ValidationError("Check 6 failed: 'decisions' is not a list")
    if len(decisions) > MAX_DECISIONS:
        raise ValidationError(f"Check 6 failed: 'decisions' has more than {MAX_DECISIONS} entries")

    # 7. action_items is a list (may be empty)
    action_items = prompt1.get("action_items", [])
    if not isinstance(action_items, list):
        raise ValidationError("Check 7 failed: 'action_items' is not a list")

    # 8. primary_category is a valid slug
    primary_category = prompt2.get("primary_category", "")
    if primary_category not in VALID_CATEGORY_SLUGS:
        raise ValidationError(f"Check 8 failed: 'primary_category' '{primary_category}' is not valid")

    # 9. urgency is a valid value
    urgency = prompt2.get("urgency", "")
    if urgency not in VALID_URGENCY:
        raise ValidationError(f"Check 9 failed: 'urgency' '{urgency}' is not valid (must be routine/notable/significant)")

    # 10. fiscal_impact is a boolean
    fiscal_impact = prompt2.get("fiscal_impact")
    if not isinstance(fiscal_impact, bool):
        raise ValidationError(f"Check 10 failed: 'fiscal_impact' must be a boolean, got {type(fiscal_impact)}")

    logger.debug("All 10 validation checks passed for item: %s", title[:60])

File: frontend/lib/pipeline/validator.ts
/**
 * Quality validator for agenda items produced by the LLM pipeline.
 * Mirrors the 10-check spec from task 5.
 * Call before insertAgendaItem; skip the item (don't crash the document) on failure.
 */

export interface AgendaItemDraft {
  title: string
  summary: string
  decisions: string[]
  primary_category: string
  urgency: string
  key_figures: {
    amounts: string[]
    vote_tallies: string[]
    dates?: string[]
    schools?: string[]
  }
  source_pdf_url: string
}

export interface ValidationResult {
  valid: boolean
  errors: string[]
}

const VALID_CATEGORIES = new Set([
  'schools-education', 'school-construction', 'budget-finance', 'transportation',
  'zoning-land-use', 'public-safety', 'policy-governance', 'equity-inclusion',
  'technology', 'community-parks', 'personnel', 'general',
])

const VALID_URGENCY = new Set(['routine', 'notable', 'significant'])

// $1M | $2.3M | $450K | $4,800,000 | $1.2B (case-insensitive suffix)
const CURRENCY_RE = /^\$[\d,]+(\.\d+)?[KMBkmb]?$/

// 5-2 | 6–1 | 4-3-0 | 7-0 (supports en-dash and three-way splits)
const VOTE_TALLY_RE = /^\d+[-–]\d+([-–]\d+)?$/

const URL_RE = /^https?:\/\/.+/

/**
 * Validate one agenda item draft before writing to the DB.
 *
 * @param item           The draft produced by the LLM
 * @param seenTitles     Mutable set of normalised titles already accepted in this meeting.
 *                       Pass the same Set instance across all items for one document so
 *                       check #8 (no duplicates within a meeting) works correctly.
 */
export function validateAgendaItem(
  item: AgendaItemDraft,
  seenTitles: Set<string>,
): ValidationResult {
  const errors: string[] = []

  // 1. Title not empty, under 100 chars
  const trimmedTitle = item.title?.trim() ?? ''
  if (trimmedTitle.length === 0) {
    errors.push('Title is empty')
  } else if (trimmedTitle.length > 100) {
    errors.push(`Title too long: ${trimmedTitle.length} chars (max 100)`)
  }

  // 2. Summary between 50 and 500 chars
  const summaryLen = item.summary?.trim().length ?? 0
  if (summaryLen < 50) {
    errors.push(`Summary too short: ${summaryLen} chars (min 50)`)
  } else if (summaryLen > 500) {
    errors.push(`Summary too long: ${summaryLen} chars (max 500)`)
  }

  // 3. primary_category must be one of the 12 valid slugs
  if (!VALID_CATEGORIES.has(item.primary_category)) {
    errors.push(`Invalid primary_category: "${item.primary_category}"`)
  }

  // 4. urgency must be routine | notable | significant
  if (!VALID_URGENCY.has(item.urgency?.toLowerCase())) {
    errors.push(`Invalid urgency: "${item.urgency}"`)
  }

  // 5. Non-routine items must have at least one decision
  if (item.urgency !== 'routine' && (item.decisions?.length ?? 0) === 0) {
    errors.push(`Non-routine item (urgency="${item.urgency}") has no decisions`)
  }

  // 6. Dollar amounts must look like currency strings
  for (const amount of item.key_figures.amounts ?? []) {
    const s = amount?.trim()
    if (s && !CURRENCY_RE.test(s)) {
      errors.push(`Invalid currency amount: "${s}"`)
    }
  }

  // 7. Vote tallies must match N-N or N-N-N pattern
  for (const tally of item.key_figures.vote_tallies ?? []) {
    const s = tally?.trim()
    if (s && !VOTE_TALLY_RE.test(s)) {
      errors.push(`Invalid vote tally: "${s}"`)
    }
  }

  // 8. No duplicate titles within the same meeting
  const normTitle = trimmedTitle.toLowerCase()
  if (normTitle && seenTitles.has(normTitle)) {
    errors.push(`Duplicate agenda item title in this meeting: "${trimmedTitle}"`)
  } else if (normTitle) {
    seenTitles.add(normTitle)
  }

  // 9. Page range — the current chunker splits on header patterns, not page boundaries,
  //    so no page-range metadata is tracked. This check is a no-op and documented here
  //    so it can be enabled if the extractor gains per-page tracking in the future.

  // 10. Source PDF URL must be a valid http(s) URL
  if (!URL_RE.test(item.source_pdf_url ?? '')) {
    errors.push(`Invalid source_pdf_url: "${item.source_pdf_url}"`)
  }

  return { valid: errors.length === 0, errors }
}
